### ISSUE TITLE
`racket-mode` and `error-display-handler` aren't playing nice

### ISSUE THREAD

**Issue by zyrolasting** [racket/racket#4131]
Related: https://github.com/greghendershott/racket-mode/issues/585

Affects multiple versions.

`racket-mode` disagrees with `exn->string` re: the value of `error-display-handler` when using REPL buffers. In `racket-mode`, this program prints to the screen and does not touch `current-error-port`. The value of this form is `"1"` normally. but `""` in `racket-mode`. The `1` is instead printed to the screen as `; 1`.

```
(parameterize ([current-error-port (open-output-string)])
  ((error-display-handler) "1" #f)
  (get-output-string (current-error-port)))
```

**What should have happened?**

If `exn->string` wants to respect the current parameterization, then it cannot assume that  `error-display-handler` will print to `current-error-port` like it does now. If it were up to me, I'd make `error-display-handler` depend on `exn->string` instead of the other way around. I'm unsure if that can happen, and the other options I'm thinking of seem icky.

RFC

**Comment by zyrolasting** [comment]
I expect this implementation of `racket/exn` would work if `default-error-display-handler` existed and was guaranteed to be the original display handler, regardless of how the runtime launched.

```
#lang racket/base

(provide exn->string)

(define (exn->string exn [port-parameter current-error-port])
  (if (exn? exn)
      (let ([out (open-output-string)])
        (parameterize ([current-error-port (open-output-string)]
                                 [current-error-display-handler default-error-display-handler])
          ((error-display-handler) (exn-message exn) exn)
          (get-output-string (port-parameter))))
      (format "~s\n" exn)))
```

**Comment by rfindler** [comment]
I don't think we can change the ways these handlers interact and behave. And yes, it is a bit fiddly to get it right, for sure.

Looks like expeditor/xrepl gets this one wrong too:

```
Welcome to Racket v8.4.0.2 [cs].
> (parameterize ([current-error-port (open-output-string)])
    ((error-display-handler) "1" #f)
    (get-output-string (current-error-port)))
1 [,bt for context]
""
```

But plain `racket` gets it right:

```
$  racket -q
Welcome to Racket v8.4.0.2 [cs].
> (parameterize ([current-error-port (open-output-string)])
  ((error-display-handler) "1" #f)
  (get-output-string (current-error-port)))
"1\n"
````

**Comment by zyrolasting** [comment]
Would you all be open to a `format-exn` that replicates the logic of the display handler without printing? That way both REPL designers and library authors don't have to surprise each other like this.

**Comment by rfindler** [comment]
Do you mean something that sets the current error port and then calls the
error display handler?

Robby

**Comment by zyrolasting** [comment]
Not exactly. IIRC, there is no option _other_ that to use `error-display-handler` if you want to format an exception. `fprintf`, `display`, `write`, etc. all accept output ports. Why not an error formatter? It can't be `exn->string` because it is too different, structurally.

The problem is that if I write a serialization library that encodes exceptions using `exn->string`, someone is going to get different serializations because of the REPL they use to perform the task.

**Comment by rfindler** [comment]
Oh -- I think that serializing an exception isn't the same as rendering it.
Should we not just use the usual serialization library to serialize
exceptions? Or does that not work?

**Comment by zyrolasting** [comment]
Maybe I should have said "encoding." My use case is to take a string-encoding of the exception (or rendering, as you put it), and put it in a machine readable document. Put another way: a human-readable inside a machine-readable. This behavior is why I can't do that with `exn->string`, particularly in unit tests I run in `racket-mode` with `C-c C-t`.

**Comment by rfindler** [comment]
When you say "this behavior" are you referring to the bug that you posted
here that is in emacs mode and expeditor/xrepl? If so, probably best that
that gets fixed? Am I still misunderstanding? (Apologies if so!)

Robby

**Comment by greghendershott** [comment]
There is a bug to fix in Racket Mode and I'll push a commit for that, as I just commented over there.

Even with that fixed:

> The problem is that if I write a serialization library that encodes exceptions using exn->string, someone is going to get different answers depending on the REPL they use.
> ...
> My use case is to take a string-encoding of the exception (or rendering, as you put it), and put it in a machine readable document. Put another way: a human-readable inside a machine-readable.

REPLs display various amounts of exception information using various styles.

- For example XREPL and Racket Mode display things prefixed with `;`, so they look like comments and more obviously differ from other output.
- Displaying `exn-message` is straightforward, but what about `continuation-mark-set->context`? And is it "trimmed" to exclude the context from the tool itself (as do Racket Mode and Dr Racket)?
- There's also the `exn:srclocs-accessor` business. Furthermore, some things duplicate messages in `exn-message` and these extra `srclocs`; some tools try to de-dupe that, others may not.
- `srcloc->string` elides useful complete paths (substituting prefixes like `<pkgs>/`) making them useless for the standard Emacs tools that know how to visit `/path/to/file.rkt:1:2` but don't know what to do about `<pkgs>/file.rkt:1:2` -- so Racket Mode prevents this elision.
- And probably more I'm overlooking.

So:

- If you really want "human readable inside machine readable" and it's OK for the "human readable" part to vary (wildly) among REPLs, then `exn->string` is probably OK for your purpose (after I fix the Racket Mode bug).

- But if you want consistency (i.e. if it's really "machine readable, period"), then I think you probably need to parameterize your own `error-display-handler`, or just directly write your own private version of an `exn->string` that doesn't use that parameter and does do exactly the sort of error content/formatting you want, independent of REPL?

**Comment by mflatt** [comment]
It seems that `exn->string` wants to be configurable via `error-display-handler`, but on the assumption that it always write to the current error port. Meanwhile, the value of `error-display-handler` isn't constrained to always write to the current error port (by specification or in practice, currently).

If I understand [the](https://github.com/greghendershott/racket-mode/issues/585#issuecomment-1015433116), the `racket-mode` change will be to write to the current error port, thus accommodating `exn->string`. The `expeditor` display handler could similarly do what it currently does only when the current error port is the original one. More generally, the solutions along this path constrain `error-display-handler` values to always write to the current error port.

I'm not sure it's a good idea to constrain `error-display-handler` values this way, but it works for the examples we've looked at so far, so maybe it's ok. In other words, we'd decide that the bug really is in `racket-mode` and `expeditor`, and we'd update the documentation for `error-display-handler` to say explicitly that a handler should write to the current error port.

**Comment by rfindler** [comment]
I think it would be bad to change the expectation that the
error-display-handler writes to the current-error-port. If there were some
explicit expectation that it writes to some specific poet that would make
it less bad, but there are definitely places where I have written code that
makes that assumption and strange things would be printed if that
expectation weren't met and debugging information would be lost.

Of course, this is all inside drr which generally isn't started via
expeditor.

Robby

**Comment by zyrolasting** [comment]
Much thanks to all!

> But if you want consistency (i.e. if it's really "machine readable, period"), then I think you probably need to parameterize your own error-display-handler, or just directly write your own private version of an exn->string that doesn't use that parameter and does do exactly the sort of error content/formatting you want, independent of REPL?

Yes, and I expect to do the latter. Parameterizing `error-display-handler` is harder to understand for a reader, because the name `exn->string` seems to imply that no such ritual is necessary. Since it does not accept a port as an argument, one cannot reason about it as a pure function.
